Add Cardano currency support

ddeath opened the pull request:
For the support of Cardano currency was required to add cardano-specific ed25519 crypto support.

The crypto code is based on [cardano-crypto](https://github.com/input-output-hk/cardano-crypto) and [cardano-sl](https://github.com/input-output-hk/cardano-sl) - official Cardano repos.

The functions are designed to work with the [cardanolite](https://www.cardanolite.com/) wallet - a lightweight wallet for Cardano based on the official [Daedalus](https://daedaluswallet.io/) wallet. A beta version of cardanolite allowing connecting with the experimental trezor firmware will be available soon.

We are NOT official developers of Cardano, however we are developers of [www.cardanolite.com](https://www.cardanolite.com)

prusnak commented:
Do not merge unless https://github.com/trezor/trezor-common/pull/148 is merged first

ddeath commented:
@prusnak there was a problem with firmware build, where it errored on `undefined reference to _sbrk`

For our testing purposes I applied fix as suggested [here](http://ww1.microchip.com/downloads/en/DeviceDoc/Frequently-Asked-Questions-4.9.3.26.txt) in point 6 and 12.

I added `-lrdimon` but it should also work with `-lnosys` not sure which one is correct or better

ddeath commented:
Will look into it. We are not using malloc directly but from build output it looks that something is:

```
arm-none-eabi-gcc -o build/firmware/firmware.elf -Os -g3 -nostdlib -std=gnu99 -Wall -Werror -Wdouble-promotion -Wpointer-arith -fno-common -mthumb -mtune=cortex-m4 -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -fsingle-precision-constant -fdata-sections -ffunction-sections -ffreestanding -fstack-protector-all -Wno-sequence-point -DGITREV=b529969 -DPRODUCTION=0 -DPYOPT=1 -T embed/firmware/memory.ld -Wl,--gc-sections -Wl,-Map=build/firmware/firmware.map -Wl,--warn-common build/firmware/embed/extmod/modtrezorconfig/modtrezorconfig.o build/firmware/embed/extmod/modtrezorconfig/norcow.o build/firmware/embed/extmod/modtrezorconfig/storage.o build/firmware/embed/extmod/modtrezorcrypto/modtrezorcrypto.o build/firmware/embed/extmod/modtrezorcrypto/crc.o build/firmware/embed/extmod/modtrezorcrypto/rand.o build/firmware/vendor/trezor-crypto/address.o build/firmware/vendor/trezor-crypto/aes/aescrypt.o build/firmware/vendor/trezor-crypto/aes/aeskey.o build/firmware/vendor/trezor-crypto/aes/aes_modes.o build/firmware/vendor/trezor-crypto/aes/aestab.o build/firmware/vendor/trezor-crypto/base58.o build/firmware/vendor/trezor-crypto/base32.o build/firmware/vendor/trezor-crypto/bignum.o build/firmware/vendor/trezor-crypto/bip32.o build/firmware/vendor/trezor-crypto/bip39.o build/firmware/vendor/trezor-crypto/blake256.o build/firmware/vendor/trezor-crypto/blake2b.o build/firmware/vendor/trezor-crypto/blake2s.o build/firmware/vendor/trezor-crypto/curves.o build/firmware/vendor/trezor-crypto/ecdsa.o build/firmware/vendor/trezor-crypto/chacha20poly1305/chacha20poly1305.o build/firmware/vendor/trezor-crypto/chacha20poly1305/chacha_merged.o build/firmware/vendor/trezor-crypto/chacha20poly1305/poly1305-donna.o build/firmware/vendor/trezor-crypto/chacha20poly1305/rfc7539.o build/firmware/vendor/trezor-crypto/ed25519-donna/curve25519-donna-32bit.o build/firmware/vendor/trezor-crypto/ed25519-donna/curve25519-donna-helpers.o build/firmware/vendor/trezor-crypto/ed25519-donna/curve25519-donna-scalarmult-base.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519-donna-32bit-tables.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519-donna-basepoint-table.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519-donna-impl-base.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519-keccak.o build/firmware/vendor/trezor-crypto/ed25519-donna/ed25519-sha3.o build/firmware/vendor/trezor-crypto/ed25519-donna/modm-donna-32bit.o build/firmware/vendor/trezor-crypto/groestl.o build/firmware/vendor/trezor-crypto/hasher.o build/firmware/vendor/trezor-crypto/hmac.o build/firmware/vendor/trezor-crypto/memzero.o build/firmware/vendor/trezor-crypto/nem.o build/firmware/vendor/trezor-crypto/nist256p1.o build/firmware/vendor/trezor-crypto/pbkdf2.o build/firmware/vendor/trezor-crypto/rand.o build/firmware/vendor/trezor-crypto/ripemd160.o build/firmware/vendor/trezor-crypto/secp256k1.o build/firmware/vendor/trezor-crypto/sha2.o build/firmware/vendor/trezor-crypto/sha3.o build/firmware/embed/extmod/modtrezorio/modtrezorio.o build/firmware/embed/extmod/modtrezorui/display.o build/firmware/embed/extmod/modtrezorui/inflate.o build/firmware/embed/extmod/modtrezorui/font_bitmap.o build/firmware/embed/extmod/modtrezorui/font_roboto_bold_20.o build/firmware/embed/extmod/modtrezorui/font_roboto_regular_20.o build/firmware/embed/extmod/modtrezorui/font_robotomono_regular_20.o build/firmware/embed/extmod/modtrezorui/modtrezorui.o build/firmware/embed/extmod/modtrezorui/trezor-qrenc/qr_encode.o build/firmware/embed/extmod/modtrezorutils/modtrezorutils.o build/firmware/embed/firmware/modutime.o build/firmware/embed/firmware/startup.o build/firmware/embed/firmware/header.o build/firmware/embed/firmware/main.o build/firmware/embed/firmware/mphalport.o build/firmware/embed/firmware/nlrthumb.o build/firmware/vendor/micropython/extmod/modubinascii.o build/firmware/vendor/micropython/extmod/moductypes.o build/firmware/vendor/micropython/extmod/moduheapq.o build/firmware/vendor/micropython/extmod/modutimeq.o build/firmware/vendor/micropython/extmod/utime_mphal.o build/firmware/vendor/micropython/lib/embed/abort_.o build/firmware/vendor/micropython/lib/mp-readline/readline.o build/firmware/vendor/micropython/lib/utils/interrupt_char.o build/firmware/vendor/micropython/lib/utils/printf.o build/firmware/vendor/micropython/lib/utils/pyexec.o build/firmware/vendor/micropython/lib/utils/stdout_helpers.o build/firmware/vendor/micropython/py/argcheck.o build/firmware/vendor/micropython/py/asmarm.o build/firmware/vendor/micropython/py/asmbase.o build/firmware/vendor/micropython/py/asmthumb.o build/firmware/vendor/micropython/py/asmx64.o build/firmware/vendor/micropython/py/asmx86.o build/firmware/vendor/micropython/py/asmxtensa.o build/firmware/vendor/micropython/py/bc.o build/firmware/vendor/micropython/py/binary.o build/firmware/vendor/micropython/py/builtinevex.o build/firmware/vendor/micropython/py/builtinhelp.o build/firmware/vendor/micropython/py/builtinimport.o build/firmware/vendor/micropython/py/compile.o build/firmware/vendor/micropython/py/emitbc.o build/firmware/vendor/micropython/py/emitcommon.o build/firmware/vendor/micropython/py/emitglue.o build/firmware/vendor/micropython/py/emitinlinethumb.o build/firmware/vendor/micropython/py/emitinlinextensa.o build/firmware/vendor/micropython/py/formatfloat.o build/firmware/vendor/micropython/py/frozenmod.o build/firmware/vendor/micropython/py/lexer.o build/firmware/vendor/micropython/py/malloc.o build/firmware/vendor/micropython/py/map.o build/firmware/vendor/micropython/py/modarray.o build/firmware/vendor/micropython/py/modbuiltins.o build/firmware/vendor/micropython/py/modgc.o build/firmware/vendor/micropython/py/modmath.o build/firmware/vendor/micropython/py/modmicropython.o build/firmware/vendor/micropython/py/modstruct.o build/firmware/vendor/micropython/py/modsys.o build/firmware/vendor/micropython/py/mpprint.o build/firmware/vendor/micropython/py/mpstate.o build/firmware/vendor/micropython/py/mpz.o build/firmware/vendor/micropython/py/nativeglue.o build/firmware/vendor/micropython/py/obj.o build/firmware/vendor/micropython/py/objarray.o build/firmware/vendor/micropython/py/objattrtuple.o build/firmware/vendor/micropython/py/objbool.o build/firmware/vendor/micropython/py/objboundmeth.o build/firmware/vendor/micropython/py/objcell.o build/firmware/vendor/micropython/py/objclosure.o build/firmware/vendor/micropython/py/objcomplex.o build/firmware/vendor/micropython/py/objdict.o build/firmware/vendor/micropython/py/objenumerate.o build/firmware/vendor/micropython/py/objexcept.o build/firmware/vendor/micropython/py/objfilter.o build/firmware/vendor/micropython/py/objfloat.o build/firmware/vendor/micropython/py/objfun.o build/firmware/vendor/micropython/py/objgenerator.o build/firmware/vendor/micropython/py/objgetitemiter.o build/firmware/vendor/micropython/py/objint_longlong.o build/firmware/vendor/micropython/py/objint_mpz.o build/firmware/vendor/micropython/py/objint.o build/firmware/vendor/micropython/py/objlist.o build/firmware/vendor/micropython/py/objmap.o build/firmware/vendor/micropython/py/objmodule.o build/firmware/vendor/micropython/py/objnamedtuple.o build/firmware/vendor/micropython/py/objnone.o build/firmware/vendor/micropython/py/objobject.o build/firmware/vendor/micropython/py/objpolyiter.o build/firmware/vendor/micropython/py/objproperty.o build/firmware/vendor/micropython/py/objrange.o build/firmware/vendor/micropython/py/objreversed.o build/firmware/vendor/micropython/py/objset.o build/firmware/vendor/micropython/py/objsingleton.o build/firmware/vendor/micropython/py/objslice.o build/firmware/vendor/micropython/py/objstr.o build/firmware/vendor/micropython/py/objstringio.o build/firmware/vendor/micropython/py/objstrunicode.o build/firmware/vendor/micropython/py/objtuple.o build/firmware/vendor/micropython/py/objtype.o build/firmware/vendor/micropython/py/objzip.o build/firmware/vendor/micropython/py/opmethods.o build/firmware/vendor/micropython/py/parse.o build/firmware/vendor/micropython/py/parsenum.o build/firmware/vendor/micropython/py/parsenumbase.o build/firmware/vendor/micropython/py/persistentcode.o build/firmware/vendor/micropython/py/qstr.o build/firmware/vendor/micropython/py/reader.o build/firmware/vendor/micropython/py/repl.o build/firmware/vendor/micropython/py/runtime_utils.o build/firmware/vendor/micropython/py/runtime.o build/firmware/vendor/micropython/py/scope.o build/firmware/vendor/micropython/py/sequence.o build/firmware/vendor/micropython/py/showbc.o build/firmware/vendor/micropython/py/smallint.o build/firmware/vendor/micropython/py/stackctrl.o build/firmware/vendor/micropython/py/stream.o build/firmware/vendor/micropython/py/unicode.o build/firmware/vendor/micropython/py/vstr.o build/firmware/vendor/micropython/py/warning.o build/firmware/vendor/micropython/ports/stm32/gccollect.o build/firmware/vendor/micropython/ports/stm32/gchelper.o build/firmware/vendor/micropython/ports/stm32/pendsv.o build/firmware/vendor/micropython/py/gc.o build/firmware/vendor/micropython/py/vm.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_can.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dac_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dac.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rng.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rtc_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rtc.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_sd.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_spi.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_sram.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_fmc.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_sdmmc.o build/firmware/vendor/micropython/lib/stm32lib/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.o build/firmware/embed/trezorhal/common.o build/firmware/embed/trezorhal/flash.o build/firmware/embed/trezorhal/mini_printf.o build/firmware/embed/trezorhal/mpu.o build/firmware/embed/trezorhal/rng.o build/firmware/embed/trezorhal/sbu.o build/firmware/embed/trezorhal/sdcard.o build/firmware/embed/trezorhal/stm32.o build/firmware/embed/trezorhal/systick.o build/firmware/embed/trezorhal/touch.o build/firmware/embed/trezorhal/usb.o build/firmware/embed/trezorhal/usbd_conf.o build/firmware/embed/trezorhal/usbd_core.o build/firmware/embed/trezorhal/usbd_ctlreq.o build/firmware/embed/trezorhal/usbd_ioreq.o build/firmware/embed/trezorhal/util.o build/firmware/embed/trezorhal/vectortable.o build/firmware/frozen_mpy.o build/firmware/embed/firmware/vendorheader.o -lc_nano -lm -lgcc
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-abs.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memcmp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memcpy.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memmove.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memset.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_copysign.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_fpclassify.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_frexp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_ldexp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_modf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_nan.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_scalbn.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sprintf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strchr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strcpy.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strlen.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strncmp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-errno.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-impure.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-svfprintf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-vfprintf_i.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sf_finite.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memchr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-freer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-mallocr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-reallocr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sbrkr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-msizer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-reent.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sbrkr.o): In function `_sbrk_r':
/build/newlib-5zwpxE/newlib-2.2.0+git20150830.5a3d536/build_nano/arm-none-eabi/armv7e-m/fpu/newlib/libc/reent/../../../../../../../newlib/libc/reent/sbrkr.c:58: undefined reference to `_sbrk'
```

prusnak commented:
You have really old gcc for arm. Please use the Docker build (`build.sh`) or install the toolchain from here: https://github.com/trezor/trezor-core/blob/master/Dockerfile#L15

ddeath commented:
It is caused by `sprintf` will replace it

ddeath commented:
Fixed, `sprintf` was in the crypto repo. It was replaced by `decitoa` implementation for convertin int to string and I removed build script modifications from this repo.

prusnak commented:
Hi! I just merged in https://github.com/trezor/trezor-crypto/pull/160

Can you please rebase on top of master?

ddeath commented:
Done, rebased and fixed linting errors

prusnak commented on embed/extmod/modtrezorcrypto/modtrezorcrypto-bip32.h:
```diff
@@ -300,6 +338,20 @@ STATIC mp_obj_t mod_trezorcrypto_HDNode_public_key(mp_obj_t self) {
+///     '''
+///     Returns public key without the first bytes of the HD node.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_HDNode_public_key_cardano(mp_obj_t self) {
```
Don't introduce new function. Just call `public_key` and strip the first byte using `[1:]` in python.

prusnak commented on embed/extmod/modtrezorcrypto/modtrezorcrypto-bip32.h:
```diff
@@ -462,11 +517,46 @@ STATIC mp_obj_t mod_trezorcrypto_bip32_from_seed(mp_obj_t seed, mp_obj_t curve_n
     { MP_ROM_QSTR(MP_QSTR_HDNode), MP_ROM_PTR(&mod_trezorcrypto_HDNode_type) },
     { MP_ROM_QSTR(MP_QSTR_deserialize), MP_ROM_PTR(&mod_trezorcrypto_bip32_deserialize_obj) },
     { MP_ROM_QSTR(MP_QSTR_from_seed), MP_ROM_PTR(&mod_trezorcrypto_bip32_from_seed_obj) },
+    { MP_ROM_QSTR(MP_QSTR_from_mnemonic_cardano),
```
Put this on one line.

prusnak commented on embed/extmod/modtrezorcrypto/modtrezorcrypto-ed25519.h:
```diff
@@ -91,6 +91,33 @@ STATIC mp_obj_t mod_trezorcrypto_ed25519_sign(size_t n_args, const mp_obj_t *arg
+///     '''
+///     Uses secret key to produce the cardano signature of message.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_ed25519_sign_cardano(mp_obj_t secret_key, mp_obj_t secret_extenion, mp_obj_t message) {
```
Rename to `mod_trezorcrypto_ed25519_sign_ext`, also there are couple of typos `secret_extenion` -> `secret_extension`.

prusnak commented on src/apps/cardano/sign_transaction.py:
```diff
@@ -0,0 +1,269 @@
+    witnesses = []
+    for index, node in enumerate(nodes):
+        # \x01\x1a\x2d\x96\x4a\x09\x58\x20 is a magic cardano prefix
+        message = b"\x01\x1a\x2d\x96\x4a\x09\x58\x20" + tx_aux_hash
```
Let's extract `b"\x01\x1a\x2d\x96\x4a\x09\x58\x20"` as CARDANO_WITNESS_MAGIC constant or something similar.

ddeath commented:
Done, CR fixes are committed.

prusnak commented:
Changes in `embed/extmod/modtrezorcrypto/*` look fine!

@jpochyla will look at `src/apps/cardano/*` and `tests/*`

ddeath commented:
Fixed CR notes and also added small number of type hints. Micropython was on version 3.4 and typing is accepted in python 3.5 so I used just `:list` instead of `List(int)`

ddeath commented:
@jpochyla are the changes sufficient or does it need some more changes?

jpochyla commented:
Merged with very minor, mostly code-style, changes. Thanks a lot!

refi93 commented:
@KingHodor Trezor supports the new derivation scheme (also referred to as V2), used by Yoroi. The "D" addresses are the old ones.
